File: app/TabRecorder.cs
using System;
using System.Diagnostics;
using System.IO;

namespace LightRecorder {

  /// <summary>
  /// Writes a browser-tab recording to disk.
  ///
  /// When the chosen source is a single tab, ffmpeg is not involved. Only the
  /// browser can capture one tab in isolation - and capture only that tab's
  /// audio, which is the whole point - so the extension encodes with
  /// MediaRecorder and streams chunks over the local bridge. Our only job is to
  /// append them as they arrive, so nothing accumulates in memory.
  /// </summary>
  internal class TabRecorder {

    public Recorder.RecState State = Recorder.RecState.Idle;
    public string OutFile;
    public long Bytes;
    public DateTime StartedAt;

    public event Action<Progress> ProgressChanged;
    public event Action<string> Failed;

    FileStream _stream;
    string _actualExt;

    public bool IsRecording {
      get { return State == Recorder.RecState.Recording || State == Recorder.RecState.Starting; }
    }

    public long ElapsedMs {
      get { return IsRecording ? ActiveMs() : 0; }
    }

    // The browser does the actual pausing; this only keeps the clock honest.
    bool _paused;
    long _pauseStarted, _pausedTicks;

    public bool IsPaused { get { return _paused; } }

    public void Pause() {
      if (State != Recorder.RecState.Recording || _paused) return;
      _paused = true;
      _pauseStarted = Stopwatch.GetTimestamp();
    }

    public void Resume() {
      if (!_paused) return;
      _pausedTicks += Stopwatch.GetTimestamp() - _pauseStarted;
      _paused = false;
    }

    long ActiveMs() {
      long paused = _pausedTicks + (_paused ? Stopwatch.GetTimestamp() - _pauseStarted : 0);
      return (long)(DateTime.UtcNow - StartedAt).TotalMilliseconds - paused * 1000 / Stopwatch.Frequency;
    }

    public bool Start(Settings s, out string error) {
      error = null;
      if (IsRecording) { error = "Already recording"; return false; }

      try {
        Directory.CreateDirectory(s.OutputDir);
        // The browser picks the container (MP4 when it can, WebM otherwise) and
        // only says which once encoding has begun. Rather than race that, the
        // file opens straight away and the extension is corrected at the end.
        OutFile = Settings.UniquePath(s.OutputDir, s.BuildFilename("mp4"));
        _stream = new FileStream(OutFile, FileMode.Create, FileAccess.Write, FileShare.Read, 1 << 16);
      } catch (Exception err) {
        error = "Could not open the output file: " + err.Message;
        return false;
      }

      Bytes = 0;
      _actualExt = null;
      _paused = false;
      _pausedTicks = 0;
      StartedAt = DateTime.UtcNow;
      State = Recorder.RecState.Recording;
      return true;
    }

    public void Write(byte[] chunk, int count) {
      FileStream s = _stream;
      if (s == null || State == Recorder.RecState.Idle || count <= 0) return;

      try {
        s.Write(chunk, 0, count);
      } catch (Exception err) {
        Action<string> f = Failed;
        if (f != null) f("Write failed: " + err.Message);
        return;
      }

      Bytes += count;
      Action<Progress> h = ProgressChanged;
      if (h != null) {
        var p = new Progress();
        p.Bytes = Bytes;
        p.TimeMs = ActiveMs();
        h(p);
      }
    }

    public void SetContainerFromMime(string mime) {
      if (string.IsNullOrEmpty(mime)) return;
      _actualExt = mime.IndexOf("mp4", StringComparison.OrdinalIgnoreCase) >= 0 ? "mp4" : "webm";
    }

    public RecordResult Stop() {
      if (State == Recorder.RecState.Idle) return null;
      State = Recorder.RecState.Stopping;

      FileStream s = _stream;
      _stream = null;
      if (s != null) {
        try { s.Flush(); } catch (Exception) { }
        try { s.Dispose(); } catch (Exception) { }
      }

      // Rename if the browser gave us WebM after we optimistically said mp4.
      if (!string.IsNullOrEmpty(_actualExt) && !string.IsNullOrEmpty(OutFile) &&
          !OutFile.EndsWith("." + _actualExt, StringComparison.OrdinalIgnoreCase)) {
        try {
          string renamed = Path.ChangeExtension(OutFile, "." + _actualExt);
          File.Move(OutFile, renamed);
          OutFile = renamed;
        } catch (Exception) { /* the file is still there under the old name */ }
      }

      var result = new RecordResult();
      result.File = OutFile;
      result.DurationMs = ActiveMs();
      result.Error = Bytes == 0 ? "Nothing was captured from the tab." : null;

      State = Recorder.RecState.Idle;
      _paused = false;
      return result;
    }
  }
}

File: legacy/src/main/tabRecorder.js
'use strict';
/**
 * Writes a browser-tab recording to disk.
 *
 * When the chosen source is a single tab, ffmpeg is not involved: only the
 * browser can capture one tab in isolation (and capture only that tab's
 * audio, which is the whole point). The extension encodes with MediaRecorder
 * and streams chunks over the local bridge; our only job is to append them to
 * a file as they arrive so nothing accumulates in memory.
 */
const { EventEmitter } = require('events');
const fs = require('fs');
const path = require('path');
const { Recorder } = require('./recorder');

class TabRecorder extends EventEmitter {
  constructor() {
    super();
    this.state = 'idle';
    this.stream = null;
    this.outFile = null;
    this.bytes = 0;
    this.startedAt = 0;
    this.actualExt = null;
  }

  get isRecording() {
    return this.state === 'recording' || this.state === 'starting';
  }

  start({ settings, container }) {
    if (this.isRecording) throw new Error('Already recording');
    fs.mkdirSync(settings.outputDir, { recursive: true });

    const ext = container === 'webm' ? 'webm' : 'mp4';
    const filename = Recorder.buildFilename(settings.filenamePattern, ext);
    this.outFile = Recorder.uniquePath(settings.outputDir, filename);

    this.stream = fs.createWriteStream(this.outFile);
    this.stream.on('error', err => this.emit('error', `Write failed: ${err.message}`));

    this.bytes = 0;
    this.actualExt = null;
    this.startedAt = Date.now();
    this.state = 'recording';
    this.emit('started', { file: this.outFile });
    return { file: this.outFile };
  }

  write(chunk) {
    if (!this.stream || this.state === 'idle') return;
    const buf = Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk);
    this.bytes += buf.length;
    this.stream.write(buf);
    this.emit('progress', {
      bytes: this.bytes,
      timeMs: Date.now() - this.startedAt,
      frames: 0, fps: 0, dropped: 0, speed: ''
    });
  }

  /**
   * The browser picks the container (MP4 when it can, WebM otherwise) and
   * only tells us once encoding has begun. Rather than race that, we open the
   * file straight away and correct the extension at the end.
   */
  setContainerFromMime(mime) {
    if (!mime) return;
    this.actualExt = /mp4/i.test(mime) ? 'mp4' : 'webm';
  }

  stop() {
    if (this.state === 'idle') return Promise.resolve(null);
    this.state = 'stopping';
    return new Promise(resolve => {
      const finish = () => {
        // Rename if the browser gave us WebM after we optimistically said mp4.
        if (this.actualExt && this.outFile && !this.outFile.toLowerCase().endsWith(`.${this.actualExt}`)) {
          const renamed = this.outFile.replace(/\.[^.]+$/, `.${this.actualExt}`);
          try { fs.renameSync(this.outFile, renamed); this.outFile = renamed; } catch (_) {}
        }
        const result = {
          file: this.outFile,
          durationMs: Date.now() - this.startedAt,
          bytes: this.bytes,
          error: this.bytes === 0 ? 'Nothing was captured from the tab.' : null
        };
        this.state = 'idle';
        this.stream = null;
        this.emit('stopped', result);
        resolve(result);
      };
      if (this.stream) this.stream.end(finish);
      else finish();
    });
  }

  status() {
    return {
      state: this.state,
      file: this.outFile,
      startedAt: this.startedAt,
      elapsedMs: this.isRecording ? Date.now() - this.startedAt : 0,
      progress: { bytes: this.bytes, timeMs: Date.now() - this.startedAt }
    };
  }
}

module.exports = { TabRecorder };
